REGRESSION(286103@main): [Debug] `fast/canvas/webgl/canvas-webgl-page-cache.html` is a flaky crash

obyknovenius opened the pull request:
#### 5ea352280424fe1276451efb0a27c03e7c70a57e
<pre>
REGRESSION(286103@main): [Debug] `fast/canvas/webgl/canvas-webgl-page-cache.html` is a flaky crash
<a href="https://bugs.webkit.org/show_bug.cgi?id=283554">https://bugs.webkit.org/show_bug.cgi?id=283554</a>

Reviewed by Anne van Kesteren.

`fast/canvas/webgl/canvas-webgl-page-cache.html` is wrongly accused.
The actual culprit is `fast/canvas/webgl/canvas-test.html` but it
crashes after reporting a failure and confusing the test runner.

Results history [1] proves this by showing that the pattern of
`canvas-webgl-page-cache.html` crashes absolutely matches the pattern of
`canvas-test.html` failures.

286103@main made `testRunner.displayAndTrackRepaints()` asynchronous. In
`canvas-test.html`, it&apos;s being called from `setInterval()`&apos;s callback
before `clearInterval()`. That can lead to a situation where the
callback is triggered again while the first execution is waiting for
`displayAndTrackRepaints()` to be resolved.

Results of failing test runs [2] prove that by showing logs reported by
the callback multiple times.

This patch fixes the test by moving `displayAndTrackRepaints()` out of
the callback. This requires wrapping the test logic into an `async`
function.

Also, similarly to 157930@main, finishing the test was simplified by
calling `finishTest()` from `webgl-test.js`. This should speed it up.

1. <a href="https://results.webkit.org/?suite=layout-tests&amp">https://results.webkit.org/?suite=layout-tests&amp</a>;suite=layout-tests&amp;test=fast%2Fcanvas%2Fwebgl%2Fcanvas-test.html&amp;test=fast%2Fcanvas%2Fwebgl%2Fcanvas-webgl-page-cache.html&amp;style=debug&amp;limit=5000
2. <a href="https://build.webkit.org/results/GTK-Linux-64-bit-Debug-Tests/287141@main%20(14983)/fast/canvas/webgl/canvas-test-actual.txt">https://build.webkit.org/results/GTK-Linux-64-bit-Debug-Tests/287141@main%20(14983)/fast/canvas/webgl/canvas-test-actual.txt</a>

* LayoutTests/fast/canvas/webgl/canvas-test.html:

Canonical link: <a href="https://commits.webkit.org/287155@main">https://commits.webkit.org/287155@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dc33e7c7b7b22e815db18e423f459736f9b98e77

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/78533 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/57578 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/31915 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/83194 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/29798 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/66729 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/5859 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/61513 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/19428 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/81600 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/51524 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/69697 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/41824 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/48870 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/25342 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/28135 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/69970 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/25715 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/84560 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/5898 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/4044 "Found 2 new test failures: fast/selectors/selection-window-inactive-stroke-color.html imported/w3c/web-platform-tests/html/browsers/the-window-object/open-close/open-features-tokenization-noreferrer.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/69738 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/6059 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/67500 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/68992 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17195 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13013 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/11445 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/5845 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/5833 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/9267 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/7620 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aproskuryakov reviewed (commented):
You appear to be on the right path here, but I don't see how `unload` would be different from `addEventListener`.

I didn't analyze this test deeply, but one theory could be that the page cache didn't work for some reason, and the b/f navigation resulted in an actual load again.

obyknovenius committed:
REGRESSION(286103@main): [Debug] `fast/canvas/webgl/canvas-webgl-page-cache.html` is a flaky crash
https://bugs.webkit.org/show_bug.cgi?id=283554

Reviewed by Anne van Kesteren.

`fast/canvas/webgl/canvas-webgl-page-cache.html` is wrongly accused.
The actual culprit is `fast/canvas/webgl/canvas-test.html` but it
crashes after reporting a failure and confusing the test runner.

Results history [1] proves this by showing that the pattern of
`canvas-webgl-page-cache.html` crashes absolutely matches the pattern of
`canvas-test.html` failures.

286103@main made `testRunner.displayAndTrackRepaints()` asynchronous. In
`canvas-test.html`, it's being called from `setInterval()`'s callback
before `clearInterval()`. That can lead to a situation where the
callback is triggered again while the first execution is waiting for
`displayAndTrackRepaints()` to be resolved.

Results of failing test runs [2] prove that by showing logs reported by
the callback multiple times.

This patch fixes the test by moving `displayAndTrackRepaints()` out of
the callback. This requires wrapping the test logic into an `async`
function.

Also, similarly to 157930@main, finishing the test was simplified by
calling `finishTest()` from `webgl-test.js`. This should speed it up.

1. https://results.webkit.org/?suite=layout-tests&suite=layout-tests&test=fast%2Fcanvas%2Fwebgl%2Fcanvas-test.html&test=fast%2Fcanvas%2Fwebgl%2Fcanvas-webgl-page-cache.html&style=debug&limit=5000
2. https://build.webkit.org/results/GTK-Linux-64-bit-Debug-Tests/287141@main%20(14983)/fast/canvas/webgl/canvas-test-actual.txt

* LayoutTests/fast/canvas/webgl/canvas-test.html:

Canonical link: https://commits.webkit.org/287155@main
